### PR TITLE
Use sanic-limiter to throttle api requests

### PR DESCRIPTION
**Pull request by pankajkoti**
Throttles API requests based on user's IP address. 
With this PR, at the moment, we are only throttling `/requests` endpoint.

closes: #230

### CODE REVIEW

**Comment by pankajkoti** [comment]
Throttling in action 

Screenshot 2023-12-26 at 5 07 54 PM

**Review by jlaneve — COMMENTED** [review]
I don’t know if these rates are permissive enough - the UI polls the API every few seconds

what’s the intent behind this? can we gate just the “submit prompt” endpoint?

**Comment by pankajkoti** [comment]
> I don’t know if these rates are permissive enough - the UI polls the API every few seconds
> 
> what’s the intent behind this? can we gate just the “submit prompt” endpoint?

yes, we can very well do route based throttling. I thought the UI would not poll more than a certain number of times per minute for the requests/request_id endpoint, no?

The context is here though for intent https://astronomer.slack.com/archives/C05QJA9LTR9/p1703159886303619 (this has been linked in the issue #230 description)

**Comment by pankajastro** [comment]
LGTM. This will have no impact on the Slack path since it does not make rest API calls, right?
https://github.com/astronomer/ask-astro/blob/main/api/ask_astro/services/questions.py#L31

**Comment by pankajkoti** [comment]
> LGTM. This will have no impact on the Slack path since it does not make rest API calls, right? https://github.com/astronomer/ask-astro/blob/main/api/ask_astro/slack/controllers/mention.py#L119

yes, correct it won't impact Slack calls. However, Slack could impose rate limits by itself based on the limits specified here https://api.slack.com/apis/rate-limits

**Review by pankajastro — APPROVED** [review]
LGTM

File: ui/src/routes/+page.server.ts [hunk 1]
```diff
@@ -1,13 +1,20 @@
...
     const requests = await fetch(`${ASK_ASTRO_API_URL}/requests`);
 
+    if (requests.status === 429) {
+      throw error(429, "Too many requests");
+    }
+
     return requests.json();
   } catch (err) {
+    if (err.status === 429) {
+      throw error(429, "Too many requests");
```
Comment (sunank200) [inline ui/src/routes/+page.server.ts:16]:
```suggestion
      throw error(429, "You have made too many requests and exceeded the rate limit. Please wait for some time before trying again.");
```

File: ui/src/routes/+page.server.ts [hunk 2]
```diff
@@ -1,13 +1,20 @@
...
-import { redirect } from "@sveltejs/kit";
+import {error, redirect} from "@sveltejs/kit";
 import type { PageServerLoad } from "./$types";
 
 export const load: PageServerLoad = async () => {
   try {
     const requests = await fetch(`${ASK_ASTRO_API_URL}/requests`);
 
+    if (requests.status === 429) {
+      throw error(429, "Too many requests");
```
Comment (sunank200) [inline ui/src/routes/+page.server.ts:10]:
```suggestion
      throw error(429, "You have made too many requests and exceeded the rate limit. Please wait for some time before trying again.");
```

File: ui/src/routes/+page.server.ts [hunk 3]
```diff
@@ -41,6 +48,11 @@ export const actions = {
       },
     });
 
+
+    if (response.status === 429) {
+      throw error(429, "Too many requests");
```
Comment (sunank200) [inline ui/src/routes/+page.server.ts:53]:
```suggestion
      throw error(429, "You have made too many requests and exceeded the rate limit. Please wait for some time before trying again.");
```

**Review by sunank200 — APPROVED** [review]

**Comment by pankajkoti** [comment]
Result after throttled requests endpoint post applying review suggestions.
Screenshot 2023-12-27 at 6 05 26 PM

**Pull request merged**